zio-http/src/main/scala/zio/http/netty/model/Conversions.scala

```diff
@@ -82,17 +82,25 @@ private[netty] object Conversions {
       },
     )
 
-  private def encodeHeaderListToNetty(headersList: List[Product2[CharSequence, CharSequence]]): HttpHeaders = {
+  private def encodeHeaderListToNetty(headers: Iterable[Header]): HttpHeaders = {
     val (exceptions, regularHeaders) =
-      headersList.partition(h => h._1.toString.contains(HeaderNames.setCookie.toString))
-    val combinedHeaders              = regularHeaders
-      .groupBy(_._1)
-      .map { case (key, tuples) =>
-        key -> tuples.map(_._2).mkString(",")
-      }
+      headers.partition(h => h.headerName.toString.contains(HeaderNames.setCookie.toString))
```

Comment:
We end up going `toString` on the header name so many times, but obviously if it's not a string, that has a cost.

I wonder if our headers should store both `String` and (potentially optimized) `CharSequence` representation?

Reply:
I have an idea that when you construct `Header.ContentType("application/json")` it's immediately encoded into `Array[Byte]`. So if you use `x match { Header.ContentType(s) => ... }` it's decoded from the bytes. i.e. eager encoding, lazy decoding. 

Such things could make `CharSequence` obsolete. For now all the `CharSequence` ceremony seems like we pay a lot for something that is so brittle, it ends up being destroyed. Maybe we should switch to strings for now and rip out a big source of complexity?

Reply:
Right now we use Netty's built-in cached HTTP header names which are `AsciiString`s. By the way, their `toString` is cached so there is no point building our own cached one on top of it. Netty's request decoder also creates `AsciiString` header names and values, and for some it uses these cached names (seems to me that only for a few default ones like `Host`).

The header parser also seems to copy the source array when creating these `AsciiStrings` so it's not just pointing to the native buffer or anything like that. 

That said I don't really see much advantage of using `CharSequence`s everywhere other than making the API more painful. Using `String`s everywhere would simplify many things.

Reply:
The `Header.Custom`, which holds the untyped headers coming from Netty, could still refer to the `CharSequence`s internally, to avoid converting them to strings if the headers are not used for anything.